File: backend/lib/hotelflux/domain/habitacion.ex
defmodule HotelFlux.Domain.Habitacion do
  @moduledoc """
  Módulo de dominio para la entidad Habitación.
  Define el esquema, cambioset, validaciones de transición de estados
  y consultas funcionales sobre colecciones de habitaciones.
  """
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key {:id, :binary_id, autogenerate: true}
  @foreign_key_type :binary_id

  # Lista de estados válidos que puede tener una habitación
  @estados_validos ~w(disponible reservada ocupada en_limpieza en_mantenimiento bloqueada)

  schema "habitaciones" do
    field :numero, :string
    field :tipo, :string
    field :piso, :integer
    field :capacidad, :integer
    field :precio_noche, :decimal
    field :estado, :string, default: "disponible"
    field :caracteristicas, :map
    field :clasificacion, :string
    field :eliminado, :boolean, default: false
    field :eliminado_en, :utc_datetime
    timestamps(type: :utc_datetime)
  end

  @doc """
  Construye un struct de Habitación con estado inicial "disponible".
  Función pura: recibe atributos, devuelve nuevo struct sin efectos secundarios.
  """
  def novo(attrs \\ %{}) do
    struct(__MODULE__, Map.merge(%{estado: "disponible"}, attrs))
  end

  @doc """
  Cambioset de Ecto para validar y filtrar atributos de una habitación.
  Validaciones: campos requeridos, tipo válido y precio_noche mayor a 0.
  """
  def changeset(habitacion, attrs) do
    habitacion
    |> cast(attrs, [:numero, :tipo, :piso, :capacidad, :precio_noche, :estado, :caracteristicas, :clasificacion])
    |> validate_required([:numero, :tipo, :piso, :capacidad, :precio_noche])
    |> validate_inclusion(:tipo, ~w(simple doble suite presidencial familiar individual))
    # Validación personalizada: precio_noche debe ser un número positivo o un Decimal mayor a 0
    |> validate_change(:precio_noche, fn :precio_noche, value ->
      cond do
        is_nil(value) -> []
        is_struct(value, Decimal) and Decimal.compare(value, Decimal.new("0")) == :gt -> []
        is_number(value) and value > 0 -> []
        true -> [precio_noche: "must be greater than 0"]
      end
    end)
  end

  alias HotelFlux.Domain.Transitions

  @doc """
  Valida si una transición de estado es permitida según la máquina de estados.
  Función pura: busca en la tabla de transiciones si existe la tupla (desde → hasta).
  Retorna {:ok, nuevo_estado} o {:error, :transicion_invalida}.
  """
  def validar_transicion(%__MODULE__{estado: estado_actual}, nuevo_estado) do
    transiciones = Transitions.tabla_habitacion()
    # Busca en la lista de tuplas {evento, desde, hasta} si la transición es válida
    case Enum.find(transiciones, fn
           {_evento, desde, hasta} -> desde == estado_actual and hasta == nuevo_estado
         end) do
      {_evento, _desde, _hasta} -> {:ok, nuevo_estado}
      nil -> {:error, :transicion_invalida}
    end
  end

  @doc """
  Cambia el estado de una habitación validando primero la transición.
  Función pura: retorna {:ok, changeset} o {:error, :transicion_invalida}.
  """
  def cambiar_estado(%__MODULE__{} = habitacion, nuevo_estado) do
    with {:ok, _} <- validar_transicion(habitacion, nuevo_estado) do
      {:ok, change(habitacion, estado: nuevo_estado)}
    end
  end

  @doc "Verifica si la habitación está disponible. Función pura — predicado."
  def disponible?(%__MODULE__{estado: "disponible"}), do: true
  def disponible?(%__MODULE__{}), do: false

  @doc "Verifica si la habitación está ocupada. Función pura — predicado."
  def ocupada?(%__MODULE__{estado: "ocupada"}), do: true
  def ocupada?(%__MODULE__{}), do: false

  @doc "Retorna la lista de estados válidos. Función pura."
  def estados_validos, do: @estados_validos

  @doc "Retorna la tabla de transiciones válidas desde Transitions. Función pura."
  def transiciones_validas, do: Transitions.tabla_habitacion()

  @doc """
  Reconstruye el estado actual de una habitación aplicando una lista de eventos de dominio en orden.
  Event Sourcing: función recursiva pura que pliega eventos sobre el struct.
  """
  def reconstruir_desde_eventos(habitacion, []), do: habitacion
  def reconstruir_desde_eventos(habitacion, [evento | resto]) do
    nueva = aplicar_evento_dominio(habitacion, evento)
    reconstruir_desde_eventos(nueva, resto)
  end

  # Aplica un evento individual al struct de la habitación
  defp aplicar_evento_dominio(hab, %{tipo: "estado_cambiado", payload: %{"estado" => estado}}) do
    %{hab | estado: estado}
  end
  defp aplicar_evento_dominio(hab, %{tipo: "precio_actualizado", payload: %{"precio_noche" => precio}}) do
    %{hab | precio_noche: Decimal.new(precio)}
  end
  # Evento desconocido: se ignora (tolerancia a eventos futuros)
  defp aplicar_evento_dominio(hab, _evento), do: hab

  @doc """
  Agrupa una lista de habitaciones por su número de piso.
  Función pura — retorna un mapa %{piso => [habitaciones]}.
  """
  def agrupar_por_piso(habitaciones) do
    agrupar_recursivo(habitaciones, %{})
  end

  # Función auxiliar recursiva de cola para agrupar por piso
  defp agrupar_recursivo([], acumulador), do: acumulador
  defp agrupar_recursivo([hab | resto], acumulador) do
    actualizado = Map.update(acumulador, hab.piso, [hab], &[hab | &1])
    agrupar_recursivo(resto, actualizado)
  end

  @doc """
  Retorna una función predicado que filtra habitaciones por estado.
  Útil para composición con `filtrar_con/2`. Función pura — HOF.
  """
  def filtrar_por_estado(estado) do
    fn %__MODULE__{estado: est} -> est == estado end
  end

  @doc """
  Filtra una lista de habitaciones aplicando múltiples predicados (AND lógico).
  Función pura — usa composición de predicados.
  """
  def filtrar_con(habitaciones, predicados) when is_list(predicados) do
    Enum.filter(habitaciones, fn hab ->
      Enum.all?(predicados, fn pred -> pred.(hab) end)
    end)
  end

  # Suma el precio_noche solo si la habitación está disponible (ingreso potencial)
  defp sumar_ingreso_potencial(acc, %__MODULE__{estado: "disponible", precio_noche: p}),
    do: Decimal.add(acc, p || Decimal.new(0))
  defp sumar_ingreso_potencial(acc, _hab), do: acc

  @doc """
  Calcula estadísticas de una lista de habitaciones: total, conteo por estado e ingresos potenciales.
  Función pura — reduce la colección a un resumen.
  """
  def calcular_estadisticas(habitaciones) do
    Enum.reduce(habitaciones, %{total: 0, por_estado: %{}, ingresos_potenciales: Decimal.new(0)}, fn hab, acc ->
      %{acc |
        total: acc.total + 1,
        por_estado: Map.update(acc.por_estado, hab.estado, 1, &(&1 + 1)),
        ingresos_potenciales: sumar_ingreso_potencial(acc.ingresos_potenciales, hab)}
    end)
  end
end

File: frontend-reception/src/domain/entidades/habitacion.ts
// ═══════════════════════════════════════════════════════════
// HotelFlux — Entidad: Habitación
// ═══════════════════════════════════════════════════════════

export type EstadoHabitacion =
  | 'disponible'
  | 'reservada'
  | 'ocupada'
  | 'en_limpieza'
  | 'en_mantenimiento'
  | 'bloqueada';

export type TipoHabitacion = 'simple' | 'doble' | 'suite' | 'presidencial';

export interface Habitacion {
  readonly id: string;
  readonly numero: string;
  readonly tipo: TipoHabitacion;
  readonly piso: number;
  readonly capacidad: number;
  readonly precio_noche: string;
  readonly estado: EstadoHabitacion;
  readonly amenidades: readonly string[];
  readonly clasificacion: string | null;
  readonly caracteristicas: Record<string, unknown> | null;
  readonly notas: string | null;
  readonly inserted_at: string;
  readonly updated_at: string;
}

export const COLOR_ESTADO: Readonly<Record<EstadoHabitacion, string>> = {
  disponible: '#10b981',
  reservada: '#3b82f6',
  ocupada: '#ef4444',
  en_limpieza: '#f59e0b',
  en_mantenimiento: '#8b5cf6',
  bloqueada: '#6b7280',
} as const;

/**
 * [TAILWIND v4] Clases de color semánticas para estados de habitación.
 *
 * Las variables CSS están definidas en index.css bajo @theme:
 *   --color-estado-disponible, --color-estado-reservada, etc.
 *
 * Usar estas clases en lugar de style={{ backgroundColor }} permite
 * que Tailwind aplique prefijos hover:, dark:, md:, etc.
 */
export const CLASE_ESTADO: Readonly<Record<EstadoHabitacion, string>> = {
  disponible: 'bg-estado-disponible',
  reservada: 'bg-estado-reservada',
  ocupada: 'bg-estado-ocupada',
  en_limpieza: 'bg-estado-limpieza',
  en_mantenimiento: 'bg-estado-mantenimiento',
  bloqueada: 'bg-estado-bloqueada',
} as const;

export const LABEL_ESTADO: Readonly<Record<EstadoHabitacion, string>> = {
  disponible: 'Disponible',
  reservada: 'Reservada',
  ocupada: 'Ocupada',
  en_limpieza: 'En Limpieza',
  en_mantenimiento: 'Mantenimiento',
  bloqueada: 'Bloqueada',
} as const;

// ─────────────────────────────────────────────────────────────────
// FUNCIONES PURAS DE DOMINIO — sin efectos secundarios
// ─────────────────────────────────────────────────────────────────

// ── Tipo de evento de dominio para Event Sourcing ──

export interface EventoHabitacion {
  readonly tipo: 'estado_cambiado' | 'precio_actualizado' | 'notas_actualizadas';
  readonly habitacion_id: string;
  readonly payload: Readonly<Record<string, unknown>>;
  readonly insertado_en: string;
}

/**
 * Reconstruye el estado de una habitación aplicando una lista de eventos.
 * FUNCIÓN RECURSIVA PURA — Event Sourcing en el cliente.
 *
 * El estado actual ES la proyección de todos los eventos anteriores.
 * Patrón: fold/reduce recursivo sobre lista inmutable.
 *
 * ## Ejemplo
 *   const estadoActual = reconstruirDesdeEventos(habitacionBase, eventos);
 */
export function reconstruirDesdeEventos(
  habitacion: Habitacion,
  eventos: readonly EventoHabitacion[],
): Habitacion {
  if (eventos.length === 0) return habitacion; // caso base

  const [evento, ...resto] = eventos;
  const nueva = aplicarEvento(habitacion, evento!); // transformación pura

  return reconstruirDesdeEventos(nueva, resto); // tail recursion
}

// Función pura: aplica un evento al struct (sin mutación)
function aplicarEvento(habitacion: Habitacion, evento: EventoHabitacion): Habitacion {
  switch (evento.tipo) {
    case 'estado_cambiado':
      return { ...habitacion, estado: evento.payload['estado'] as EstadoHabitacion };
    case 'precio_actualizado':
      return { ...habitacion, precio_noche: String(evento.payload['precio_noche']) };
    case 'notas_actualizadas':
      return { ...habitacion, notas: String(evento.payload['notas'] ?? '') };
    default:
      return habitacion; // evento desconocido: estado sin cambios
  }
}

/**
 * HOF: filtrarPorEstado — función que retorna función (currying).
 * Retorna un predicado para usar con Array.filter o en streams RxJS.
 *
 * ## Ejemplo
 *   const soloDisponibles = filtrarPorEstado('disponible');
 *   habitaciones.filter(soloDisponibles);
 */
export function filtrarPorEstado(
  estado: EstadoHabitacion,
): (habitacion: Habitacion) => boolean {
  // Closure: captura `estado` en la función retornada
  return (habitacion: Habitacion) => habitacion.estado === estado;
}

/**
 * HOF: filtrarCon — acepta múltiples predicados como argumentos.
 * Una habitación pasa si TODOS los predicados retornan true.
 * HOF: recibe array de funciones, aplica composición lógica.
 *
 * ## Ejemplo
 *   const disponiblesGrandes = filtrarCon(
 *     habitaciones,
 *     [filtrarPorEstado('disponible'), h => h.capacidad >= 2]
 *   );
 */
export function filtrarCon(
  habitaciones: readonly Habitacion[],
  predicados: ReadonlyArray<(h: Habitacion) => boolean>,
): readonly Habitacion[] {
  return habitaciones.filter((hab) =>
    predicados.reduce((pasa: boolean, predicado) => pasa && predicado(hab), true),
  );
}

/**
 * HOF: transformarCon — aplica una lista de transformaciones en secuencia.
 * Composición de funciones puras sobre la habitación.
 * HOF: recibe funciones como argumentos.
 *
 * ## Ejemplo
 *   const transformada = transformarCon(hab, [
 *     h => ({ ...h, estado: 'disponible' }),
 *     h => ({ ...h, notas: 'Limpieza completada' }),
 *   ]);
 */
export function transformarCon(
  habitacion: Habitacion,
  transformaciones: ReadonlyArray<(h: Habitacion) => Habitacion>,
): Habitacion {
  return transformaciones.reduce((hab, transform) => transform(hab), habitacion);
}

/**
 * Calcula estadísticas de ocupación. FUNCIÓN PURA (fold/reduce).
 * HOF implícita: usa reduce con función acumuladora inline.
 *
 * ## Ejemplo
 *   const stats = calcularEstadisticas(habitaciones);
 *   // → { disponibles: 5, ocupadas: 8, porcentaje: 61.5 }
 */
export function calcularEstadisticas(
  habitaciones: readonly Habitacion[],
): Readonly<Record<EstadoHabitacion, number> & { total: number; porcentajeOcupacion: number }> {
  const conteo = habitaciones.reduce(
    (acc, hab) => ({
      ...acc,
      [hab.estado]: (acc[hab.estado] ?? 0) + 1,
    }),
    {} as Record<EstadoHabitacion, number>,
  );

  const total = habitaciones.length;
  const ocupadas = conteo['ocupada'] ?? 0;

  return {
    disponible: conteo['disponible'] ?? 0,
    reservada: conteo['reservada'] ?? 0,
    ocupada: ocupadas,
    en_limpieza: conteo['en_limpieza'] ?? 0,
    en_mantenimiento: conteo['en_mantenimiento'] ?? 0,
    bloqueada: conteo['bloqueada'] ?? 0,
    total,
    porcentajeOcupacion: total > 0 ? Math.round((ocupadas / total) * 100) : 0,
  };
}

/**
 * Agrupa habitaciones por piso. FUNCIÓN RECURSIVA PURA.
 * Construye un ReadonlyMap<piso, habitaciones[]> de forma funcional.
 *
 * ## Ejemplo
 *   const mapa = agruparPorPiso(habitaciones);
 *   mapa.get(1) // → [hab101, hab102, ...]
 */
export function agruparPorPiso(
  habitaciones: readonly Habitacion[],
): ReadonlyMap<number, readonly Habitacion[]> {
  return agruparRecursivo(habitaciones, new Map<number, Habitacion[]>());
}

// Recursión de cola: construye el mapa piso → habitaciones
function agruparRecursivo(
  habitaciones: readonly Habitacion[],
  acc: Map<number, Habitacion[]>,
): ReadonlyMap<number, readonly Habitacion[]> {
  if (habitaciones.length === 0) return acc; // caso base

  const [hab, ...resto] = habitaciones;
  const pisoActual = acc.get(hab!.piso) ?? [];
  acc.set(hab!.piso, [...pisoActual, hab!]); // spread: sin mutar el array previo

  return agruparRecursivo(resto, acc); // tail call
}

/**
 * HOF: predicado de prioridad — retorna función de comparación.
 * Útil para ordenar habitaciones según criterio configurable.
 *
 * ## Ejemplo
 *   habitaciones.sort(porPrioridad('en_mantenimiento'));
 */
export function porPrioridad(
  estadoPrioritario: EstadoHabitacion,
): (a: Habitacion, b: Habitacion) => number {
  // Función de comparación que captura estadoPrioritario (closure)
  return (a: Habitacion, b: Habitacion) => {
    if (a.estado === estadoPrioritario && b.estado !== estadoPrioritario) return -1;
    if (b.estado === estadoPrioritario && a.estado !== estadoPrioritario) return 1;
    return a.numero.localeCompare(b.numero);
  };
}
